packages/nextjs/app/page.tsx

```diff
@@ -19,6 +39,39 @@ const Home: NextPage = () => {
     }
   }
 
+  //format address for smaller screen size
+  function formatAddress(address: string) {
+    if (!address) {
+      return "Resolving...";
+    }
+    return `${address.substring(0, 5)}...${address.substring(address.length - 4)}`;
+  }
+
+  //creates a list of checked in builders addresses from the event history of the contract
+  //uses Set to make sure there are no duplicated on re-render
+  useEffect(() => {
+    if (!isReadingEventLoading && eventHistory && isConnectedOptimism) {
+      const currentBuildersSet = new Set(checkedInBuilders);
+      const builders = eventHistory
+        .map(e => e.args.builder)
+        .filter(builder => builder !== undefined && !currentBuildersSet.has(builder)) as string[];
+      if (builders.length > 0) {
+        setCheckedInBuilders(currentBuilders => [...currentBuilders, ...builders]);
+      }
+    }
+  }, [isReadingEventLoading, eventHistory, checkedInBuilders, isConnectedOptimism]);
```

Comment:
We actually don't need separate `checkedInBuilders` state and also we could remove this whole `useEffect` and do something like this : 

```ts
const buildersSet = new Set(eventHistory?.map(e => e.args.builder).filter(Boolean));
const checkedInBuilders = Array.from(buildersSet);
```
`.filter(Boolean)` will remove undefined and null values in above code 

Since `eventHistory` is internally stored using `useState` in `useScaffoldEventHistory` the `checkedInBuilders` will be updated as soon as it changes.

Reply:
Trying to get my head around this - the code is much cleaner that's for sure!